## Read / Contig Alignment Foundations

### Summary
When manipulating any type of Reads or Contig data, a common initial use case is the alignment of this data against a reference genome. For RNA-seq, for example, this is handled in KBase with the Bowtie tool from the Tuxedo suite. This works well for aligning RNA-seq results in a single sample, but in most cases, users will have several samples, and will need to view and manage results for each. We also need to expand this toolkit process to handle the general case of alignments of Reads and Contigs, not just for a single type, and improve the existing types and provide data <-> file functions for interacting with the data.


### Story / Description

The expected userâ€™s workflow is: 
 - User starts an alignment job, or, possibly several related jobs in a single batch with different input files but the same parameters.
 - Job(s) creates one or more alignment outputs with accompanying report files (SAM/BAM files).
 - User views, manages, and relates statistics from each report.
 - User views the alignment of Reads and Contigs against the target genome.

This project builds on the existing RNA-seq tools. The goal is to generalize that tool and give users the option of multiple standard read/contig alignment tools (Bowtie, bwa, etc.) that all provide similar functionality. Then, with standardized outputs, this would allow us to standardize viewers and reporting.

To support this kind of workflow, this project may depend on batch or parallel SDK job execution for running jobs, generalized data types for representing alignments, UI elements for dealing with the reports (and possibly static images generated by the tools). Ultimately, it should make use of a genome browser (backed by some dynamic service) or track viewer, though this could get deferred until a later time.

Risks: For running the jobs, there are few risks - this would build on the pattern that the RNA-seq tools have started. Creating generalized outputs might be risky, but we can adopt community standard formats (SAM, BAM, etc.) and build Workspace types based on those (reusing or refactoring the RNA-seq results as necessary). The alignment viewer is the biggest risk, but can be path toward including a genome browser in KBase.

